### ISSUE TITLE
Design Review 2019-01-23 21:00 UTC (Americas)

### ISSUE THREAD

**Issue by mrjoro** [ampproject/amphtml#20113]
Time: [2019-01-23 21:00 UTC](https://www.timeanddate.com/worldclock/meeting.html?year=2019&month=01&day=23&iv=0) ([add to Google Calendar](http://www.google.com/calendar/event?action=TEMPLATE&text=AMP%20Project%20Design%20Review&dates=20190123T210000Z/20190123T220000Z&details=https%3A%2F%2Fhangouts.google.com%2Fhangouts%2F_%2Fgoogle.com%2Famp-eng-desrev))
Location: [Video conference via Google Hangouts](https://hangouts.google.com/hangouts/_/google.com/amp-eng-desrev)

The AMP Project holds weekly engineering [design reviews](https://github.com/ampproject/amphtml/blob/master/contributing/design-reviews.md).  **We encourage everyone in the community to participate in these design reviews.**

If you are interested in bringing your design to design review, read the [design review documentation](https://github.com/ampproject/amphtml/blob/master/contributing/design-reviews.md) and add a link to your design doc by the Monday before your design review.

When attending a design review please read through the designs *before* the design review starts.  This allows us to spend more time on discussion of the design.

We rotate our design review between times that work better for different parts of the world as described in our [design review documentation](https://github.com/ampproject/amphtml/blob/master/contributing/design-reviews.md), but you are welcome to attend any design review.  If you cannot make any of the design reviews but have a design to discuss please let mrjoro@ know on [Slack](https://github.com/ampproject/amphtml/blob/master/CONTRIBUTING.md#discussion-channels) and we will find a time that works for you.

**Comment by torch2424** [comment]
I would like to present an update to amp-experiment to support Third Party AB testing providers. Intent to implement: #20225

**Comment by cpapazian** [comment]
I'd like to discuss https://github.com/ampproject/amphtml/issues/20122, which addresses some shortcomings in the current `<amp-fit-text>` component.

**Comment by mrjoro** [comment]
*amp-experiment v2*
- what changes are required on the client?
  - if we upgrade to the service worker solution it wouldn't (the final document would already have the mutations applied)
- JSON format standard?  no, it's something we came up with
- everything in the JSON is currently implementable in the current amp-experiment?
  - e.g. if you have two anchor tags, have them with different paths
  - feels a bit like "CSS inside out"
  - are there performance benefits to doing this rather than swapping out the elements?
  - we want to make it easy to make dynamic experiments on the fly so e.g. third-party experiment providers can provide WYSIWYG editors; a convenience; JSON config would be server-side included so it's a different flow than the page author modifying the doc in place with different configs; cache would get the version with the server-side code
    - another example: 3 lines where they want to test a different order; you could use amp-experiment to create all of these different mutations, hand-coded or generated, but the generated HTML would be tricky to provide; but the third-party provider already has to generate something (the JSON) so why can't they generate the HTML for amp-experiment instead?  understanding this tradeoff (e.g. perhaps the size is a lot bigger) would be helpful to decide what to do
  - when there's a single variant using amp-experiment may be straightforward, but it could get more complicated
  - @dvoytenko: investigate performance more; would also be helpful to have an actual experiment config (e.g. existing Optimizely config)
  - are there cases where the existing amp-experiment generates a document that's too large?  a comparison would help us understand
  - could we generate the HTML for the variants from the JSON?  client has to choose the variant
- for third-party providers like Optimizely they have a wide variety of sites they have to deal with so making assumptions about the layout is difficult
  - changes people typically make in Optimizely: moving elements above/below others, edit innerHTML directly, custom JS (for more complex cases), insert changes
    - allowing innerHTML mutations through a JSON config would be slow; for innerHTML we were planning on using mustache templates, but that still has to go through sanitization/purification before it ends up on the page so it's still not as fast as it could be; template instantiation or static HTML template could help
- where would boundaries be for users to understand what will work in AMP vs. non-AMP?
  - user experience design is still being worked on; potentially just adding an "AMP experiment eligible" icon
- @davidstrauss: it feels like a bit of a reboot of XSLT :)
- doing stuff in the cache is an option, but we need to keep in mind there are a lot of different caches, not just the Google cache
- how does this interact with web packaging?  the server-side includes would have to happen before the web packaging (it's up to the publishers to do this)
  - perhaps web packaging needs a way for different parts to be independently signed
  - this points to how important it is for a client-side only solution to work
- weights
  - should it be total or a percentage?  this needs to be ironed out
  - what does Optimizely use?  uses an integer weight (sums up to 10000); is there an industry standard?  not sure
- rearrangement of DOM nodes is performant
  - enforce a sort of flex-box concept on it?
  - could we make it like amp-experiment v1 with some added order changes?  or perhaps two models--one for DOM mutations, one for style changes
- we need to make sure to consider integration effort of third-party experiment platforms with whatever solution we end up with
- is there a second vendor we can include in designing this feature to avoid over-fitting the design?  we'd like to do that; @jeffjose should look into this
- does the latency effect of experiments potentially skew the experiment results (e.g. it makes it slower so users react differently?)
  - Optimizely measures execution time for the experiments; most of the time isn't due to DOM manipulations but to boilerplate code, so it's probably the same for all variations fo an experiment
- @aghassemi: should we build this on type of amp-script instead?  it already does some load-time mutations; @torchh will follow up with @kristoferbaxter
- is there a fallback for really slow devices?  there isn't a good way of undoing mutations that were already done, so you need to figure out the mutations up front before applying them
  - or can we opt people who are on really slow networks out of the experiments
  - this could also be problematic with web packaging
